source/fe/fe_poly_tensor.cc

```diff
@@ -67,17 +69,26 @@ namespace internal
         const unsigned int dim = 2;
         // const unsigned int spacedim = 2;
 
-        for (unsigned int f = GeometryInfo<dim>::faces_per_cell / 2;
-             f < GeometryInfo<dim>::faces_per_cell;
-             ++f)
+        const CellId this_cell_id = cell->id();
+
+        for (unsigned int f = 0; f < GeometryInfo<dim>::faces_per_cell; ++f)
           {
             typename dealii::Triangulation<dim, spacedim>::face_iterator face =
               cell->face(f);
             if (!face->at_boundary())
               {
                 const unsigned int nn = cell->neighbor_face_no(f);
-
-                if (nn < GeometryInfo<dim>::faces_per_cell / 2)
+                const typename dealii::Triangulation<dim,
+                                                     spacedim>::cell_iterator
+                             neighbor_cell_at_face = cell->neighbor(f);
+                const CellId neigbor_cell_id = neighbor_cell_at_face->id();
+
+                // if (nn < GeometryInfo<dim>::faces_per_cell / 2)
+                // Only fix sign if the orientation is opposite and if it has
+                // not been fixed yet.
+                // NOTE: This assumes that we iterate through cells in
+                // ascending cell_id-order
+                if (((nn + f) % 2 == 0) && this_cell_id < neigbor_cell_id)
```

Comment:
Do you mean a loop over all cells? In that case, the assumption is wrong. Roughly speaking, the order in which cells are traversed is the order in which the cells were created. So, if you start from a mesh of two cells, and you now have a mesh with 2x4 cells (globally refined) you may first visit the 4 children of the first coarse mesh cell, or the 4 children of the second coarse mesh cell, depending on what steps you took to arrive at the mesh with the 2x4 cells.

Reply:
I see. My observation was that (as was probably intended also before my change) that one should only correct a sign once when checking neighbor orientations. If we have a cell whose face dof signs have been corrected one must not do so again. The approach before unfortunately fails on some mesh configurations.

Reply:
OK, from the tests I see that (among others) fe/fe_values_view_09.cc is broken. That is likely because of that cell_id_order

Reply:
Right. What you need is a "tie breaker" that for a given face decides whether one or the other adjacent cell needs to flip a sign. You can't rely on the iteration order of cells (i.e., the `cell->active_cell_index()`), but the order provided by `cell_id` is stable and also stable for parallel triangulations. So, the approach should be something like this: 
* figure out whether for the current face a sign needs to be flipped
* flip the sign on that cell that has `this_cell_id < neighbor_cell_id`.

I don't understand the first part of the condition, but the second part looks correct. I just can't quite make the connection to the comment above -- we don't iterate over cells in ascending cell_id order, but I also don't think the condition actually relies on this.

Reply:
@bangerth The first part of the condition essentially says that when two faces with opposite normals meet then change the sign in the cell with the id that is higher. Does your comment mean that this is consistent? This is how I would understand that.

Reply:
After thinking about it: The tie breaker is correct :-)